File: tree/145.c
/**
 * post order iteration traversal is quite hard, try recursion first...
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Return an array of size *returnSize.
 * Note: The returned array must be malloced, assume caller calls free().
 */
void postO(struct TreeNode* root, int* returnSize, int** returnArr)
{
    if(root==NULL) return;
    postO(root->left, returnSize, returnArr);
    postO(root->right, returnSize, returnArr);
    *returnArr = realloc(*returnArr, sizeof(int)*((*returnSize)+1));
    (*returnArr)[(*returnSize)++] = root->val;
}

int* postorderTraversal(struct TreeNode* root, int* returnSize) {
    int** returnArr = (int**)malloc(sizeof(int*));
    *returnArr = (int*)malloc(sizeof(int)); //注意要给数组的第一个值分配地址 才能realloc
    *returnSize = 0;
    postO(root, returnSize, returnArr);
    return *returnArr;
}

File: tree/145.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> ret = new ArrayList<>();
        List<TreeNode> stack = new ArrayList<>();
        while (true) {
            if (root != null) {
                ret.add(0, root.val);
                if (root.left != null) {
                    stack.add(root.left);
                }
                if (root.right != null) {
                    stack.add(root.right);
                }
            }
            if (stack.size() == 0) {
                break;
            }
            root = stack.remove(stack.size()-1);
        }
        return ret;
    }
}
